File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists:
            return None
        
        merged_arr = []
        for l in lists:
            while l:
                merged_arr.append(l.val)
                l = l.next
        
        if not merged_arr:
            return None
        
        merged_arr.sort()
        sentinel = ListNode()
        merged_LL = ListNode(merged_arr[0])
        sentinel.next = merged_LL
        for elem in merged_arr[1:]:
            merged_LL.next = ListNode(elem)
            merged_LL = merged_LL.next
        
        return sentinel.next

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        auto cmp = [](ListNode* a, ListNode* b) {return a -> val > b -> val;};
        
        priority_queue<ListNode*, vector<ListNode*>, decltype(cmp)> pq(cmp);
        
        for (auto list: lists){
            if (list) pq.push(list);
        }
        
        ListNode* dummy = new ListNode();
        ListNode* tail = dummy;
        while (!pq.empty()){
            auto curr = pq.top(); pq.pop();
            tail -> next = curr;
            tail = tail -> next;
            if (curr -> next) pq.push(curr -> next);
        }
        
        return dummy -> next;
    }
};